### DISCUSSION TITLE
Trimming Microsoft compiler support matrix

### DISCUSSION THREAD

**Discussion by mwichmann** [SCons/scons#4392] [General]
Microsoft now supports "Visual Studio", including the compilers it runs, for a 10 year period, five years of full support and five years of extended support (no new features, security fixes only, etc.).  Earlier on these periods were shorter. SCons has an understanding of compilers going all the way back to Visual Studio 6.0, which is out of support for nearly 20 years.

This is a prod to start a discussion whether SCons should actually clean out mention of some or all of the "out of support" versions. 

Here's the product table in the same form as found on Wikipedia:

| Product name              | Code name    | Release date                      | Version  number | Latest update version    | Latest update date | Support ends                                                    | Supported  .NET Framework  (no add-on) | Supported .NET  (formerly .NET Core)  (no add-on) |   |
|---------------------------|--------------|-----------------------------------|-----------------|--------------------------|--------------------|-----------------------------------------------------------------|----------------------------------------|---------------------------------------------------|---|
| Visual Studio 97          | Boston       | March 19, 1997                    | 5.0             | Service Pack 3           | December 4, 1997   | Old version, no longer maintained: June 30, 2003                | —                                      | —                                                 |   |
| Visual Studio 6.0         | Aspen        | September 2, 1998                 | 6.0             | Service Pack 6           | March 29, 2004     | Old version, no longer maintained: September 30, 2005           | —                                      | —                                                 |   |
| Visual Studio .NET (2002) | Rainier      | February 13, 2002                 | 7.0             | Service Pack 1           | March 8, 2005      | Old version, no longer maintained: July 14, 2009                | 1.0                                    | —                                                 |   |
| Visual Studio .NET 2003   | Everett      | April 24, 2003                    | 7.1             | Service Pack 1           | August 15, 2006    | Old version, no longer maintained: October 8, 2013              | 1.1                                    | —                                                 |   |
| Visual Studio 2005        | Whidbey      | November 7, 2005                  | 8.0             | Service Pack 1           | December 15, 2006  | Old version, no longer maintained: April 12, 2016               | 2.0                                    | —                                                 |   |
| Visual Studio 2008        | Orcas        | November 19, 2007                 | 9.0             | Service Pack 1           | August 11, 2008    | Old version, no longer maintained: April 10, 2018               | 2.0, 3.0, 3.5                          | —                                                 |   |
| Visual Studio 2010        | Dev10        | April 12, 2010                    | 10.0            | Service Pack 1           | March 10, 2011     | Old version, no longer maintained: July 14, 2020                | 2.0–4.0                                | —                                                 |   |
| Visual Studio 2012        | Dev11        | September 12, 2012                | 11.0            | Update 5                 | August 24, 2015    | Old version, no longer maintained: January 10, 2023             | 2.0–4.5                                | —                                                 |   |
| Visual Studio 2013        | Dev12        | October 17, 2013                  | 12.0            | Update 5                 | July 20, 2015      | Older version, yet still maintained: April 9, 2024              | 2.0–4.5.1                              | —                                                 |   |
| Visual Studio 2015        | Dev14        | July 20, 2015                     | 14.0            | Update 3                 | June 27, 2016      | Older version, yet still maintained: October 14, 2025           | 2.0–4.6.1                              | 1.0                                               |   |
| Visual Studio 2017        | Dev15        | March 7, 2017                     | 15.0            | 15.9.55                  | June 13, 2023      | Older version, yet still maintained: April 13, 2027             | 3.5–4.7.2                              | 1.0, 1.1, 2.0, 2.1                                |   |
| Visual Studio 2019        | Dev16        | April 2, 2019                     | 16.0            | 16.11.29                 | August 7, 2023     | Older version, yet still maintained: April 10, 2029             | 3.5–4.8                                | 2.1, 2.2, 3.0, 3.1, 5.0                           |   |
| Visual Studio 2022        | Dev17        | November 8, 2021                  | 17.0            | 17.6.5                   | July 11, 2023      | Current stable version:  January 13, 2032                       | 3.5, 4.6–4.8.1                         | 2.1, 3.1, 5.0–7.0                                 |   |

**Comment by acmorrow** [comment]
It seems reasonable to me to drop the out of support ones. I'd suggest that if possible for SCons to detect attempts to use such out of date toolchains and issue an error message stating that support has been dropped, and to include in that error message the last version of SCons which supported them. That way, if you do need for some reason to build something with an ancient toolchain and you try to use a too-new version of SCons, you don't have to go searching: it just tells you what SCons version you can downgrade to and regain support.

**Comment by bdbaddog** [comment]
Probably need a deprecation cycle?

**Reply by acmorrow** [reply]
I'd argue you don't need a deprecation cycle. If you just pulled support for the no-longer-maintained VS codes right now, the next release of SCons would drop support for MS tools versions released before Oct 2013. That's almost a decade ago. And if you needed to use a ten year old toolset that has left support, that's ok - just use any older SCons release.

**Comment by bdbaddog** [comment]
Also there's two ways to get a specific version of MSVC:

1. Request it via MSVC_VERSION
2. Only have that version installed on your system/VM

So we'd have to detect which one was selected by above and issue deprecation warning if in the no longer supported.

Lemme float a poll on twitter/x

**Comment by mwichmann** [comment]
Maybe a middle ground?  If 4.6 comes out in the next month or two, it could include in the release note  that the following release (which presumably would land like March/April 2024, would no longer support msvc prior to (some-version).  That's a kind of deprecation... and I guess a note in the table here: https://scons.org/doc/production/HTML/scons-man.html#cv-MSVC_VERSION

**Reply by bdbaddog** [reply]
@mwichmann - that sounds workable.  Next major release won't be "quick" so any users who had an issue with such would have an opportunity and notice to let us know.

**Comment by bdbaddog** [comment]
Adding blurb here:

The goal is not to stop supporting old versions for which we already have code (because unfortunately I've seen some really old msvc versions still in use in the wild), but rather to stop working on them as they are unlikely to change at this point.
Then use develop for relatively (yeah I know) modern msvc's.
This should yield two benefits:
1 - The number of VMs and testing needed to be complete on new changes should shrink significantly
2 - Hopefully simplify the implementation if we know we can grab the info from vswhere.
